### ISSUE TITLE
question about clip text loss

### ISSUE THREAD

**Issue by xiaoyao3302** [littlepure2333/MindBridge#6]
Hi, many thanks for your wonderful work and code! 

I'm now working on fMRI decoding and during my experiments running your code, I noticed that the training loss of loss_clip_text keeps decreasing while the validation loss of loss_clip_text decreased for several epochs but kept increasing until stable. But the cosine_sim_text and mse_text kept decreasing. (The loss in terms of the image feature alignment seems to be normal.) I'm quite confused about the situation. I wonder if you ever come across similar problems before, and I also wonder if I can get some insights about the situation from you. 

Thanks again!

**Comment by littlepure2333** [comment]
Hi, thanks for your interest in our work! Yes, the situation is similar on my end, except that "cosine_sim_text" is increasing (though I believe it is a typo in your description). 

This is due to differences between the loss objective (soft_clip_loss) and other metrics (cosine_sim, mse). In practice, we mainly use "cosine_sim" to assess performance on the validation dataset.

Hope this helps : )

**Comment by xiaoyao3302** [comment]
Hi, thanks for the reply!
Actually, it is not a typo, I do notice that the loss_clip_text keeps increasing, while the cosine_sim keeps increasing, which I think is quite interesting and I'm afraid that I cannot explain on my own. I know that the better the model is trained, the higher cosine_sim will be, as it indicates the alignment performance. But the increment of loss_clip_text is quite weird, and I thought the increase in loss_clip_text and the increase in cosine_text should be mutually exclusive. But it does happen, which has bothered me for a long time. Besides, the loss value and the cosine value in terms of the image seem to be normal.
I was wondering if you could provide me with some insights.
![loss1](https://github.com/user-attachments/assets/71033a38-f4b0-4197-9580-2179da107355)
![loss2](https://github.com/user-attachments/assets/b19fb76a-9574-45e7-babe-1d624115b0be)

**Comment by littlepure2333** [comment]
Hi, thank you for pointing that out! I have a small thought that might be related to this phenomenon:
In this diffusion model, the length of the text embedding is fixed to 77 tokens (as is the case with many diffusion models that use the same CLIP text encoder, such as Stable Diffusion). As a result, when the text prompt is short, the text tokens are zero-padded to 77, resulting that the ground truth text embedding contains many zeros. While the predicted text embedding is most likely to have all non-zero values, when calculating the loss, this could lead to some mismatch due to the unaligned token lengths. On the other hand, for images, there is no padding on the tokens, which might explain why the image loss behaves normally.

Hope this helps! Feel free to share your thoughts as well :)

**Comment by xiaoyao3302** [comment]
Thanks for the insight! I think that really makes sense!

**Comment by xiaoyao3302** [comment]
Thanks for the insight! I think that really makes sense!
I used to doubt that that is due to the distribution of the NSD dataset, e.g. the textual features extracted from different stimuli might be very similar [1], I wonder if that would also lead to the increase of the loss_clip_text. What do you think?

[1] Spurious reconstruction from brain activity

**Comment by littlepure2333** [comment]
It may suggest that the text embedding prediction is most likely overfitting to the training set.
